---
title: "SRA Datasets"
author: "Nate Olson"
date: "September 20, 2014"
output: html_document
runtime: shiny
---
able of next generation sequencing datasets for use in evaluating performance of variant calling pipelines.  These datasets selected met the following criteria, single sample sequnced that has been multiple times, this can include being sequenced at multiple sequencing centers, on multiple platforms, or at the sample center using the same platform multiple times.  While closed reference genomes are available for each of the strains, no validated variant call sets exist for these datasets.

```{r, echo=FALSE,message=FALSE}
# loading required libraries
library(stringr)
library(reshape2)
library(plyr)
library(ggplot2)
library(knitr)
```

```{r, echo =FALSE}
#loading files
files <- grep("csv",list.files(path = "sra_list"), value = T)
sra_df <- ldply(str_c("sra_list/",files, sep = ""),read.csv,stringsAsFactors=F,header=T)

# removing NA columns
sra_df <- sra_df[,colSums(is.na(sra_df))<nrow(sra_df)]

# removing non-whole genome sequencing datasets
sra_df <- sra_df[!(sra_df$LibraryStrategy %in% c("MeDIP-Seq","ChIP-Seq")),]
sra_df <- sra_df[sra_df$LibrarySelection != "PCR",]

# Excluding Ecoli MV straings
sra_df <- sra_df[!grepl("MV",sra_df$ScientificName),]

# removing columns not needed for summary table
sra_df <- subset(sra_df, select = -c(ReleaseDate, LoadDate, size_MB, spots, bases, spots_with_mates, avgLength, AssemblyName, Experiment, SRAStudy, BioProject, Study_Pubmed_id, Tumor,TaxID, Submission, Consent, RunHash, ReadHash, LibrarySource, InsertSize, InsertDev,BioSample,LibraryName, LibraryStrategy, LibrarySelection, ProjectID, Platform,LibraryLayout,SampleName))
sra_df$Dataset <- str_c("[",sra_df$Run,"]","(",sra_df$download_path,")", sep = "")
sra_df$Run <- NULL; sra_df$download_path <- NULL

```

```{r echo=FALSE}
 seq_data <- function(dataset) {
     shinyApp(
         ui = fluidPage(responsive = FALSE,
                        fluidRow(style = "padding-bottom: 20px;",
                                 column(4, selectInput('org', 'Organism', choices = unique(dataset$ScientificName))),
                                 column(4, selectInput('center', 'Sequencing Center', choices =  unique(dataset$ScientificName))),
                                 column(4, selectInput('platform', 'Sequencing Platform', choices = unique(dataset$Model))),
                                 selected=names(dataset)[[3]]),
           plotOutput(sra_barplot)                        
    ),
    server = function(input, output) {
             #Combine the selected variables into a new data frame
             selectedData <- reactive({
                 dataset[dataset$ScientificName == input$org,]
                 })
             
             sra_barplot <- renderPlot({
                 ggplot(selectedData) + geom_bar(aes(x = Model)) + theme_bw() + theme(axis.text.x = element_text(angle = 90))
                 })
             }
         )
     }
```

```{r, echo = FALSE}
shinyApp(
  
  ui = fluidPage(
    selectInput("region", "Region:", 
                choices = colnames(WorldPhones)),
    plotOutput("phonePlot")
  ),
  
  server = function(input, output) {
    output$phonePlot <- renderPlot({
      barplot(WorldPhones[,input$region]*1000, 
              ylab = "Number of Telephones", xlab = "Year")
    })
  },
  
  options = list(height = 500)
)
```

Note the use of the `height` parameter to determine how much vertical space the embedded application should occupy.

You can also use the `shinyApp` function to define an application inline rather then in an external directory.

In all of R code chunks above the `echo = FALSE` attribute is used. This is to prevent the R code within the chunk from rendering in the document alongside the Shiny components.


```{r echo=FALSE}
seq_data(sra_df)
```

